File: protocol-generator/typescript/src/capability/scope.ts
import { type EcfValue } from "../codec/ecf-value.js";
import { Ecf } from "../model/index.js";
import * as Paths from "./paths.js";

/**
 * Which §5.2 matcher a grant dimension uses (0.8.1, F40). {@link Scope.matches}
 * takes it as a required argument — no default — so a new call site cannot silently
 * inherit the wrong matcher, which is exactly the F40 defect.
 */
export type ScopeKind = "id" | "path";

/**
 * §5.2 id-scope match (0.8.1, F40): literal comparison with exactly two wildcard
 * forms — bare `*` and a trailing `/*` segment-prefix. None of the §5.4 path
 * transforms apply — no leading-slash universal reading, no interior peer-wildcard, no
 * peer-relative qualification — so a pattern carrying path syntax is matched as a
 * literal string: a non-match, never a fault.
 */
export function matchesIdPattern(value: string, pattern: string): boolean {
  if (pattern === "*") {
    return true;
  }
  if (pattern.endsWith("/*")) {
    return value.startsWith(pattern.slice(0, -1));
  }
  return value === pattern;
}

/**
 * A grant scope dimension (V7 §3.6): `{include, exclude?}`. Both
 * `system/capability/path-scope` (handlers, resources) and
 * `system/capability/id-scope` (operations, peers) share this shape, but §5.2
 * matches each **by its scope type** (0.8.1, F40) — see {@link Scope.matches}.
 */
export class Scope {
  constructor(
    readonly include: readonly string[],
    readonly exclude: readonly string[] | null,
  ) {}

  static readonly empty = new Scope([], null);

  toEcf(): EcfValue {
    return Ecf.map(
      ["include", Ecf.array(this.include.map((p) => Ecf.text(p)))],
      ["exclude", this.exclude === null ? null : Ecf.array(this.exclude.map((p) => Ecf.text(p)))],
    );
  }

  static fromEcf(value: EcfValue): Scope {
    const include = Ecf.asArray(Ecf.require(value, "include")).map((v) => Ecf.asText(v));
    const excludeField = Ecf.field(value, "exclude");
    const exclude = excludeField === null ? null : Ecf.asArray(excludeField).map((v) => Ecf.asText(v));
    return new Scope(include, exclude);
  }

  /**
   * True if `value` is included and not excluded by this scope (§5.2
   * `matches_scope`). `kind` selects the matcher by scope type: `"path"`
   * (handlers, resources) canonicalizes both sides; `"id"` (operations, peers)
   * compares literally per the 0.8.1 id-scope grammar. The two MUST NOT be
   * interchanged.
   */
  matches(value: string, localPeerId: string, kind: ScopeKind): boolean {
    // 0.8.2.21 — an unmatchable exclude DENIES rather than carving out nothing, and
    // SCOPED TO PATH-SCOPE at 0.8.2.24 (N2/N3). §5.2's exclude loop tests the sentinel
    // INSIDE `if dimension_type == "system/capability/path-scope"`, and §5.4's rule is
    // likewise "a capability carrying an unmatchable PATH-SCOPE pattern is INVALID … It
    // does NOT reach `operations` or `peers` [MUST]".
    //
    // This guard used to sit OUTSIDE the type dispatch, transcribing §5.2's loop before
    // that loop grew its type test — which ran an id pattern through the §5.4 transforms
    // purely to classify it and then DENIED THE WHOLE DIMENSION on a property unrelated
    // to whether the exclude carves anything out: an `operations` exclude of `*/apply`,
    // an ordinary namespaced operation name, path-canonicalizes to the sentinel and
    // denied every operation. Over-denial, invisible on any well-formed grant.
    //
    // The id arm reaches the literal matcher below unguarded, which is correct: under the
    // id-scope grammar every non-`*` pattern is a literal, and a literal is never
    // structurally unmatchable, so there is nothing here for the sentinel to detect.
    if (kind === "path" && this.exclude !== null && Paths.excludeIsUnmatchable(this.exclude, localPeerId)) {
      return false;
    }
    const canonicalValue = kind === "path" ? Paths.canonicalize(value, localPeerId) : value;
    const covers = (pattern: string): boolean =>
      kind === "id"
        ? matchesIdPattern(value, pattern)
        : Paths.matchesPattern(canonicalValue, Paths.canonicalize(pattern, localPeerId));

    let matched = false;
    for (const pattern of this.include) {
      if (covers(pattern)) {
        matched = true;
        break;
      }
    }
    if (!matched) {
      return false;
    }

    if (this.exclude !== null) {
      for (const pattern of this.exclude) {
        if (covers(pattern)) {
          return false;
        }
      }
    }
    return true;
  }
}

File: protocol-generator/csharp/src/EntityCore.Protocol/Capability/Scope.cs
using EntityCore.Protocol.Codec;
using EntityCore.Protocol.Model;

namespace EntityCore.Protocol.Capability;

/// <summary>
/// Which §5.2 matcher a grant dimension uses (0.8.1, F40). <see cref="Scope.Matches"/>
/// takes it as a required argument — no default — so a new call site cannot silently
/// inherit the wrong matcher, which is exactly the F40 defect.
/// </summary>
internal enum ScopeKind
{
    /// <summary><c>operations</c>, <c>peers</c> — <c>system/capability/id-scope</c>.</summary>
    Id,

    /// <summary><c>handlers</c>, <c>resources</c> — <c>system/capability/path-scope</c>.</summary>
    Path,
}

/// <summary>
/// A grant scope dimension (V7 §3.6): <c>{include, exclude?}</c>. Both
/// <c>system/capability/path-scope</c> (handlers, resources) and
/// <c>system/capability/id-scope</c> (operations, peers) share this shape, but §5.2
/// matches each <em>by its scope type</em> (0.8.1, F40) — see <see cref="Matches"/>.
/// </summary>
internal sealed record Scope(IReadOnlyList<string> Include, IReadOnlyList<string>? Exclude)
{
    public static readonly Scope Empty = new(System.Array.Empty<string>(), null);

    public EcfValue ToEcf() => Ecf.Map(
        ("include", Ecf.Array(Include.Select(Ecf.Text))),
        ("exclude", Exclude is null ? null : Ecf.Array(Exclude.Select(Ecf.Text))));

    public static Scope FromEcf(EcfValue value)
    {
        IReadOnlyList<string> include = Ecf.AsArray(Ecf.Require(value, "include")).Select(Ecf.AsText).ToList();
        EcfValue? exclude = Ecf.Field(value, "exclude");
        IReadOnlyList<string>? excludes = exclude is null
            ? null
            : Ecf.AsArray(exclude).Select(Ecf.AsText).ToList();
        return new Scope(include, excludes);
    }

    /// <summary>
    /// §5.2 id-scope match (0.8.1, F40): literal comparison with exactly two wildcard
    /// forms — bare <c>*</c> and a trailing <c>/*</c> segment-prefix. None of the §5.4
    /// path transforms apply, so a pattern carrying path syntax is matched as a literal
    /// string: a non-match, never a fault.
    /// </summary>
    internal static bool MatchesIdPattern(string value, string pattern)
    {
        if (pattern == "*")
        {
            return true;
        }
        if (pattern.Length >= 2 && pattern.EndsWith("/*", StringComparison.Ordinal))
        {
            return value.StartsWith(pattern[..^1], StringComparison.Ordinal);
        }
        return value == pattern;
    }

    /// <summary>
    /// True if <paramref name="value"/> is included and not excluded by this scope
    /// (§5.2 <c>matches_scope</c>). <paramref name="kind"/> selects the matcher by scope
    /// type: <c>Path</c> (handlers, resources) canonicalizes both sides; <c>Id</c>
    /// (operations, peers) compares literally. The two MUST NOT be interchanged.
    /// </summary>
    public bool Matches(string value, string localPeerId, ScopeKind kind)
    {
        // 0.8.2.21 — an unmatchable exclude DENIES rather than carving out nothing, and
        // SCOPED TO PATH-SCOPE at 0.8.2.24 (N2/N3). §5.2's exclude loop tests the sentinel
        // INSIDE `if dimension_type == "system/capability/path-scope"`, and §5.4's rule is
        // likewise "a capability carrying an unmatchable PATH-SCOPE pattern is INVALID …
        // It does NOT reach `operations` or `peers` [MUST]".
        //
        // This guard used to sit OUTSIDE the type dispatch, transcribing §5.2's loop
        // before that loop grew its type test — which ran an id pattern through the §5.4
        // transforms purely to classify it and then DENIED THE WHOLE DIMENSION on a
        // property unrelated to whether the exclude carves anything out: an
        // <c>operations</c> exclude of <c>*/apply</c>, an ordinary namespaced operation
        // name, path-canonicalizes to the sentinel and denied every operation.
        // Over-denial, invisible on any well-formed grant.
        //
        // The id arm reaches the literal matcher below unguarded, which is correct: under
        // the id-scope grammar every non-`*` pattern is a literal, and a literal is never
        // structurally unmatchable, so there is nothing here for the sentinel to detect.
        if (kind == ScopeKind.Path && Exclude is not null && Paths.ExcludeIsUnmatchable(Exclude, localPeerId))
        {
            return false;
        }
        string canonicalValue = kind == ScopeKind.Path ? Paths.Canonicalize(value, localPeerId) : value;

        bool Covers(string pattern) => kind == ScopeKind.Id
            ? MatchesIdPattern(value, pattern)
            : Paths.MatchesPattern(canonicalValue, Paths.Canonicalize(pattern, localPeerId));

        bool matched = false;
        foreach (string pattern in Include)
        {
            if (Covers(pattern))
            {
                matched = true;
                break;
            }
        }
        if (!matched)
        {
            return false;
        }

        if (Exclude is not null)
        {
            foreach (string pattern in Exclude)
            {
                if (Covers(pattern))
                {
                    return false;
                }
            }
        }
        return true;
    }
}
